pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,425 @@
+		localEndpointSlice, ok := localEndpointSlicesMap[derivedSuffix]
+		if !ok {
+			toDeleteSuffixes = append(toDeleteSuffixes, derivedSuffix)
+		} else if needUpdate(localEndpointSlice, derivedEndpointSlice, clusterName, derivedService) {
+			toUpdateSuffixes = append(toUpdateSuffixes, derivedSuffix)
+		}
+	}
+
+	for localSuffix := range localEndpointSlicesMap {
+		if _, ok := derivedEndpointSlicesMap[localSuffix]; !ok {
+			toAddSuffixes = append(toAddSuffixes, localSuffix)
+		}
+	}
+
+	return
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) updateEndpointSlices(
+	ctx context.Context, toUpdateSuffixes []string, derivedEndpointSlicesMap,
+	localEndpointSlicesMap map[string]*discoveryv1.EndpointSlice,
+	derivedService *corev1.Service,
+) error {
+	if len(toUpdateSuffixes) == 0 {
+		return nil
+	}
+
+	for _, suffix := range toUpdateSuffixes {
+		localEndpointSlice := localEndpointSlicesMap[suffix]
+		derivedEndpointSlice := derivedEndpointSlicesMap[suffix]
+
+		updateDerivedEndpointSlice(derivedEndpointSlice, localEndpointSlice, r.clusterName, derivedService)
+		if err := r.Client.Update(ctx, derivedEndpointSlice); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func updateDerivedEndpointSlice(
+	derivedEndpointSlice, localEndpointSlice *discoveryv1.EndpointSlice,
+	clusterName string, derivedService *corev1.Service,
+) {
+	derivedEndpointSlice.OwnerReferences = []metav1.OwnerReference{{
+		APIVersion:         derivedService.APIVersion,
+		Kind:               derivedService.Kind,
+		Name:               derivedService.Name,
+		UID:                derivedService.UID,
+		Controller:         ptr.To(true),
+		BlockOwnerDeletion: ptr.To(true),
+	}}
+
+	derivedEndpointSlice.Labels = maps.Clone(derivedService.Labels)
+	if derivedEndpointSlice.Labels == nil {
+		derivedEndpointSlice.Labels = map[string]string{}
+	}
+	derivedEndpointSlice.Labels[mcsapiv1alpha1.LabelServiceName] = localEndpointSlice.Labels[discoveryv1.LabelServiceName]
+	derivedEndpointSlice.Labels[discoveryv1.LabelServiceName] = derivedService.Name
+	derivedEndpointSlice.Labels[mcsapiv1alpha1.LabelSourceCluster] = clusterName
+	derivedEndpointSlice.Labels[discoveryv1.LabelManagedBy] = endpointSliceLocalMCSAPIControllerName
+
+	derivedEndpointSlice.Annotations = maps.Clone(derivedService.Annotations)
+	derivedEndpointSlice.AddressType = localEndpointSlice.AddressType
+
+	// We need to deep copy the slices as those struct are full of pointers
+	derivedEndpointSlice.Endpoints = make([]discoveryv1.Endpoint, 0, len(localEndpointSlice.Endpoints))
+	for _, endpoint := range localEndpointSlice.Endpoints {
+		derivedEndpointSlice.Endpoints = append(derivedEndpointSlice.Endpoints, *endpoint.DeepCopy())
+	}
+	derivedEndpointSlice.Ports = make([]discoveryv1.EndpointPort, 0, len(localEndpointSlice.Ports))
+	for _, port := range localEndpointSlice.Ports {
+		derivedEndpointSlice.Ports = append(derivedEndpointSlice.Ports, *port.DeepCopy())
+	}
```

Comment:
A shallow copy should be equally fine here as well, given that the resulting object is only potentially mutated during the `Update` call, and we don't access the local endpoints object at that point anymore (controller runtime already returns a deepcopy AFAIR, differently from plain informers/listers).